File: Lakshay Garg/Leetcode/Sorting/Bag of Tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& arr, int p) {
        int count=0,i=0,j=arr.size()-1,ans=0;
        sort(arr.begin(),arr.end());
        while(i<=j){
            if(arr[i]<=p){
                count++;
                ans=max(count,ans);
                p-=arr[i];
                i++;

            }
            else{
                if(count>=1){
                    p+=arr[j];
                    j--;
                    count--;
                }
                else break;
            }
        }
        return ans;
    }
};

File: Anya_Seth/Sorting/Bag of Tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int i = 0;
        int j = tokens.length - 1;
        int score = 0;
        int maxScore = 0;
        while(i <= j){
            if(tokens[i] <= power){
               power -= tokens[i];
               score++;
               i++;
            }
            else if(score >= 1){
                power += tokens[j];
                score--;
                j--;
            }
            else{
                return 0;
            }
            maxScore = Math.max(score, maxScore);
        }
        return maxScore;
    }
}
